### PR TITLE
introduce order.stake_amount

### LINKED ISSUE #8906: adjust_trade_position()'s stake_amount handled differently between backtesting and dry-run/live while using leverage?
## Describe your environment

  * Operating system: Docker's value -> Debian GNU/Linux 11 (bullseye)
  * Python Version: Docker's value -> 3.11.4
  * CCXT version: Docker's value -> 3.1.44
  * Freqtrade Version: Docker's value -> 2023.6
  
## Your question

I am dry-running  and backtesting a strategy with `position_adjustment_enable = True`. I found out that `stake_amount` returned from  `adjust_trade_position()` during *dry-run* has become very high, even higher than wallet balance. Eventually the bot used whole wallet balance to increase the position.

However, during *backtesting* the same pair and timerange, this isn't the behaviour. The bot calculates the `stake_amount` as per intended. Here's a real life example (reproducible code below):

My DCA calculation is:
```
stake_amount = 2 * previous_stake_amount
```

Both backtesting and dry-run have same initial stake_amount of $10, leverage 10x, wallet balance of $50:

in backtesting
```
previous_stake_amount is: 9.982392999999998
after DCA, current position was increased by stake_amount of: 19.964785999999997

which is the intended behaviour where stake_amount = 2 * 9.982392999999998 = 19.964785999999997
```

however in dry-run:
```
previous_stake_amount is: 99.99028000000001
after DCA, current position (supposedly) increased by stake_amount of: 199.98056000000003
which is greater than wallet balance.

and this is big discrepancy against backtesting calculation
```

I noticed during `dry-run`, the calculation factored in `leverage` (in my case is 10x) while `backtesting` does not.

Hope you can understand the situation. I am confused why dry run calculates it differently and in theory could be in live mode too. IMO calculation in backtesting is the intended behaviour and accurate too.

Some perhaps relevant information
```
Exchange: Binance
Mode: Futures (isolated)
Pair: StaticPairList, WAVES/USDT:USDT only
--timerange 20230714-
"dry_run_wallet": 50 (same as in backtest too)
stake_amount: 10
leverage: 10
```

tldr; why dry-run handles `stake_amount` wrongly by using leverage?

Thank you and let me know if you need more info.

Here is the reproducible code (do not run in LIVE).
```python
from freqtrade.strategy import IStrategy
from pandas import DataFrame
from pandas_ta import sma
from freqtrade.persistence import Trade
from datetime import datetime
from typing import Optional
import logging

logger = logging.getLogger(__name__)

class Reaper(IStrategy):
    INTERFACE_VERSION: int = 3
    can_short = True

    # ROI table:
    minimal_roi = {
        "0": 0.1
    }

    # Stoploss:
    stoploss = -0.5

    max_open_trades = 1
    timeframe = '5m'
    startup_candle_count: int = 12
    position_adjustment_enable = True

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[:, f"fast_sma"] = sma(dataframe['close'], 6)
        dataframe.loc[:, f"slow_sma"] = sma(dataframe['close'], 12)
        return dataframe
    
    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # DON'T RUN IN LIVE, purposely made to be losing
        dataframe.loc[
            (
                dataframe['fast_sma'] < dataframe['slow_sma']
            ),
            ['enter_long', 'enter_tag']
        ] = (1, 'sma_crossed_over')
        
        dataframe.loc[
            (
                dataframe['fast_sma'] > dataframe['slow_sma']
            ),
            ['enter_short', 'enter_tag']
        ] = (1, 'sma_crossed_below')
        
        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        return dataframe

    def adjust_trade_position(self, trade: Trade, current_time: datetime,
                              current_rate: float, current_profit: float,
                              min_stake: Optional[float], max_stake: float,
                              current_entry_rate: float, current_exit_rate: float,
                              current_entry_profit: float, current_exit_profit: float,
                              **kwargs) -> Optional[float]:
        
        if current_profit > -0.01:
            return None
        
        filled_entries = trade.select_filled_orders(trade.entry_side)
        try:
            # This returns last order stake size.
            previous_stake_amount = filled_entries[-1].cost
            # double down previous entry
            stake_amount = 2 * previous_stake_amount

            if self.wallets:
                available_balance = self.wallets.get_free("USDT")
                
            logger.info(f"##### log for debugging stake_amount became very high during dry-run")
            logger.info(f"##### trade id: {trade.id}, prev_stake_amount: {previous_stake_amount}, stake_amount: {stake_amount}, wallet_balance: {available_balance}")
            
            return stake_amount
        except Exception as exception:
            return None

        return None

    def leverage(self, pair: str, current_time: datetime, current_rate: float,
                 proposed_leverage: float, max_leverage: float, entry_tag: Optional[str], side: str,
                 **kwargs) -> float:
        return 10
    
    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
                            proposed_stake: float, min_stake: Optional[float], max_stake: float,
                            leverage: float, entry_tag: Optional[str], side: str,
                            **kwargs) -> float:
        return 10

```

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary
using order.cost is not reliable to use for adjustment amounts in futures - as cost is a value that's comming directly from the exchange, and may be in contracts.
Instead, we'll be adding `order.stake_amount` - which will contain the actual stake amount used for that particular order.

Closes #8906

## Quick changelog

- fix cost assignment in backtesting (backtesting and dry will match - but live might still be different)
- add order.stake_amount
- change code sample

### CODE REVIEW

**Comment by batcandoi27** [comment]
there is a confirmation regarding this same issue when live. here: 

https://github.com/hippocritical/strategies/issues/3

after completing you should edit the sample code for dca at freqtrade.io thanks /    
'''
stake_amount = filled_entries[0].cost
'''

by the way, the crossover adjustment function can consider adding a parameter : False = entry and adjust each candle only once, True = It is possible to buy and correct multiple times in the same candle ! if I can set false it will help the backtest  and live results to be similar!

**Comment by xmatthias** [comment]
> after completing you should edit the sample code for dca at freqtrade.io thanks 

Already part of this PR :) 

> by the way, the crossover adjustment function can consider adding a parameter : False = entry and adjust each candle only once, True = It is possible to buy and correct multiple times in the same candle ! if I can set false it will help the backtest and live results to be similar!

doesn't make sense on it's own / in the context of this PR -

**Comment by batcandoi27** [comment]
after trying to replace .cost with initial_stake = filled_buys[0].stake_amount, the below error appears when dry (backtest is not error). I don't know where is the bug?? 
''''2023-08-05 05:06:42 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.7', state='RUNNING'
2023-08-05 05:07:27 - stra - INFO - Error occured while trying to get stake amount for COMP/USDT:USDT: Parent instance <Order at 0x7ff0b55005d0> is not bound to a Session; lazy load operation of attribute '_trade_live' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
2023-08-05 05:07:43 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.7', state='RUNNING'
2023-08-05 05:08:30 - stra - INFO - Error occured while trying to get stake amount for BTC/USDT:USDT: Parent instance <Order at 0x7ff0d2b33b90> is not bound to a Session; lazy load operation of attribute '_trade_live' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
2023-08-05 05:08:46 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.7', state='RUNNING'
2023-08-05 05:09:47 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.7', state='RUNNING' '''

**Comment by xmatthias** [comment]
@batcandoi27 that's a nice catch - will be fixed in develop shortly.
please retry with the latest develop version.

**Comment by batcandoi27** [comment]
> @batcandoi27 that's a nice catch - will be fixed in develop shortly. please retry with the latest develop version.

I run the version 2023.8-dev - 2069abe3 and I don't see the error anymore! And the position adjustments work perfectly! Thank you for the quick fix!

**Comment by tomashinM** [comment]
Still have this lazy load error sometimes on live when adjusting. (freqtrade 2023.10-dev)

**Comment by bw0x7n** [comment]
issue reappeared even with using freqtrade 2024.11-dev this is the full stacktrace : 

```
2024-12-11 12:13:15 INFO    freqtrade.freqtradebot - DCA order open, will wait for resolution: Trade(id=483, pair=JANET/USDT, amount=1019363.90000000, is_short=False, leverage=1.0, open_rate=0.00030014, open_since=2024-11-24 02:54:10)
2024-12-11 12:13:16 ERROR   freqtrade.strategy.strategy_wrapper - Unexpected error Parent instance <Order at 0x7f20606bfbe0> is not bound to a Session; lazy load operation of attribute '_trade_live' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3) calling <bound method emaband_ema5_sell0_97_buy1_03.adjust_trade_position of <emaband_ema5_sell0_97_buy1_03.emaband_ema5_sell0_97_buy1_03 object at 0x7f205a5b0f70>>
Traceback (most recent call last):
  File "/root/immbot/freqtrade2/freqtrade/strategy/strategy_wrapper.py", line 29, in wrapper
    return f(*args, **kwargs)
  File "/root/immbot/freqtrade2/user_data/strategies/livebotpools/emaband_ema5_sell0_97_buy1_03.py", line 212, in adjust_trade_position
    total_spent = sum(order.stake_amount for order in filled_buys)
  File "/root/immbot/freqtrade2/user_data/strategies/livebotpools/emaband_ema5_sell0_97_buy1_03.py", line 212, in <genexpr>
    total_spent = sum(order.stake_amount for order in filled_buys)
  File "/root/immbot/freqtrade2/freqtrade/persistence/trade_model.py", line 175, in stake_amount
    return self.safe_amount * self.safe_price / self.trade.leverage
  File "/root/immbot/freqtrade2/freqtrade/persistence/trade_model.py", line 170, in trade
    return self._trade_bt or self._trade_live
  File "/root/immbot/freqtrade2/.venv/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 566, in __get__
    return self.impl.get(state, dict_)  # type: ignore[no-any-return]
  File "/root/immbot/freqtrade2/.venv/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 1086, in get
    value = self._fire_loader_callables(state, key, passive)
  File "/root/immbot/freqtrade2/.venv/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 1121, in _fire_loader_callables
    return self.callable_(state, passive)
  File "/root/immbot/freqtrade2/.venv/lib/python3.10/site-packages/sqlalchemy/orm/strategies.py", line 922, in _load_for_state
    raise orm_exc.DetachedInstanceError(
sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <Order at 0x7f20606bfbe0> is not bound to a Session; lazy load operation of attribute '_trade_live' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
2024-12-11 12:13:16 ERROR   freqtrade.strategy.strategy_wrapper - Unexpected error Parent instance <Order at 0x7f205bb2c760> is not bound to a Session; lazy load operation of attribute '_trade_live' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3) calling <bound method emaband_ema5_sell0_97_buy1_03.adjust_trade_position of <emaband_ema5_sell0_97_buy1_03.emaband_ema5_sell0_97_buy1_03 object at 0x7f205a5b0f70>>
Traceback (most recent call last):
```

and this is the adjust trade position code i've used : 

```

    def adjust_trade_position(self, trade: Trade, current_time: datetime,
                              current_rate: float, current_profit: float,
                              min_stake: Optional[float], max_stake: float,
                              current_entry_rate: float, current_exit_rate: float,
                              current_entry_profit: float, current_exit_profit: float,
                              **kwargs) -> Optional[float]:
        # Level-based parameters
        min_profit_threshold = self.profit_threshold  # Level-based entry threshold
        max_level_increase = 0.25  # Increase in stake for each additional buy

        # Time-based parameters
        max_buys_per_day = 2  # Max 2 buys per 24 hours
        cooldown_period = self.cooldown_period_h  # 12-hour cooldown between each buy

        # Check level-based condition first
        if current_profit > min_profit_threshold:
            return None

        dataframe, _ = self.dp.get_analyzed_dataframe(trade.pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        previous_candle = dataframe.iloc[-2].squeeze()

        if last_candle['close'] < previous_candle['close']:
            return None

        filled_buys = trade.select_filled_orders('buy')
        count_of_buys = len(filled_buys)
        total_spent = sum(order.stake_amount for order in filled_buys)
    

        # Calculate initial stake amount
        initial_stake_amount = filled_buys[0].stake_amount if filled_buys else min_stake
        stake_amount = initial_stake_amount * (1 + (count_of_buys * max_level_increase))

        # Check if adding the next stake amount would exceed the max balance cap
        if (total_spent + stake_amount) > self.max_balance_cap:
            print(f"Cap reached for {trade.pair}: Total spent {total_spent} + Next stake {stake_amount} exceeds cap of {self.max_balance_cap}")
            return None  # Skip buy if cap exceeded

        # Time-based condition  
        if count_of_buys > 0:
            last_order = filled_buys[-1]
            if last_order.order_filled_date is not None:
                tz = last_order.order_filled_date.tzinfo
                current_time_with_tz = current_time.replace(tzinfo=tz)
                time_difference = current_time_with_tz - last_order.order_filled_date

                # Apply cooldown logic
                if time_difference.total_seconds() < cooldown_period:
                    return None  # Cooldown period not over, skip the buy

        # If all conditions are met, return the calculated stake amount
        return stake_amount


```

**Comment by xmatthias** [comment]
@bw0x7n that's a different problem to start with. This is a 2 year old  pull request - not an issue you can add your problems onto.

if you think you've discovered an issue - it's surely something **new** - so please open a new issue for it.
